page-tracking/src/hw_mem_map.rs

```diff
@@ -146,6 +147,96 @@ impl HwMemRegion {
         let pages = self.size / PageSize::Size4k as u64;
         self.base.checked_add_pages(pages).unwrap()
     }
+
+    /// Returns the page size if this has been partitioned, or None if it hasn't
+    pub fn page_size(&self) -> Option<PageSize> {
+        self.page_size
+    }
+
+    /// break a `HwMemRegion` into multiple `HwMemRegion`s that can all fit into vm pages of the same size
+    pub fn partition(&self, page_size: PageSize) -> RegionVec {
+        let mut array_vec = RegionVec::default();
+
+        // degenerate case
+        if self.size == 0 {
+            return array_vec;
+        }
+
+        // current size to small to hold a page, go to a smaller page
+        if self.size < page_size as u64 {
+            return self.partition(page_size.down_size().expect("bad page size"));
+        }
+
+        // We have at lease one page. If this is aligned we are done
+        if page_size.is_aligned(self.base.bits()) && page_size.is_aligned(self.size) {
+            let mut new_reg = *self;
+            new_reg.page_size = Some(page_size);
+            array_vec.push(new_reg);
+            return array_vec;
+        }
+
+        // if this isn't aligned and couldn't contain a page size (i.e it is less than 2 pages long)
+        // down_size the page and try again
+        if page_size.num_pages(self.size) < 2 && !page_size.is_aligned(self.base.bits()) {
+            return self.partition(page_size.down_size().expect("bad page size"));
```

Comment:
Not sure of whether it's right in this context, but usually we avoid panics.

Reply:
I'll switch to unwrap, which also panics but at least doesn't also give a confusing message.

Reply:
Ah OK...what I meant is that we usually propagate the error upstream (by using say a Result<> type), and let the caller deal with it. It's possible that in this case, we ***are***  always expecting callers to pass us correct parameters, so unwrap() is a way to discover that something was amiss.